File: components/src/dynamo/trtllm/request_handlers/batched_egress.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Flag-gated cross-request batched egress for the TRT-LLM decode worker.

Root cause (DYN-3703): every response token is pulled out of a *per-request*
Python async generator and crosses the Python->Rust boundary individually,
through the pyo3-async bridge, on ONE GIL-bound event loop (~37 us/token on a
96-core node). Per decode iteration at batch B that is B * 37 us on a single
lane, which eventually can't keep up with the engine and back-pressures prefill.

This module batches egress ACROSS requests: all of a step's ready responses are
handed to Rust in ONE bridge crossing, then Rust demuxes them back to each
request's response stream by id. It is the analogue of TRT-LLM's
``handle_for_ipc_batched`` (base_worker.py:860), which is exactly how
``trtllm serve`` keeps its egress cost from scaling with batch.

Design goal: reuse ``handler.generate`` UNCHANGED, so cancellation, error
mapping, disagg prefill/decode, and metrics behave identically to the
per-request path. The only new work is (a) a per-request feeder that tags each
chunk with the request id and pushes it onto a shared queue, and (b) a single
``drain`` generator that coalesces whatever is ready into one yield.

Enable with ``DYN_TRTLLM_BATCHED_EGRESS=1`` (default OFF -> no behavior change).

Pairs with the Rust ``Endpoint.serve_endpoint_batched(submit, drain, ...)`` +
``PythonBatchedEgressEngine`` added in the same patch.
"""

import asyncio
import logging
import os
from typing import Any, AsyncGenerator, Optional


def batched_egress_enabled() -> bool:
    """True when DYN_TRTLLM_BATCHED_EGRESS selects the batched path."""
    return os.environ.get("DYN_TRTLLM_BATCHED_EGRESS", "0").lower() not in (
        "0",
        "",
        "false",
        "no",
    )


# Sentinel pushed after a request's generator completes so the Rust demux can
# close that request's response stream. Sent over the queue as the chunk value;
# the drain converts it to ``None`` on the wire so Rust does not have to know
# about this Python object.
_DONE = object()


class BatchedEgress:
    """Owns the shared egress queue, the per-request feeders, and the drain.

    One instance per worker. ``submit`` is scheduled once per request by the
    Rust engine; ``drain`` is driven once by the Rust engine's single forwarder.
    """

    def __init__(self, handler: Any) -> None:
        self._handler = handler
        # Unbounded: back-pressure is applied downstream by each request's Rust
        # mpsc channel (RESPONSE_CHANNEL_DEPTH). Keeping the shared queue
        # unbounded avoids one slow client stalling the whole worker's drain.
        self._q: "asyncio.Queue[tuple[str, Any]]" = asyncio.Queue()

    async def submit(self, request: dict, context: Any) -> None:
        """Run one request's existing generator, tag each chunk with the Dynamo
        request id, and push it onto the shared queue.

        Reuses ``handler.generate`` verbatim, so all per-request semantics are
        preserved. Any exception is surfaced as a tagged error frame so the Rust
        side can map it exactly like the per-request path does.
        """
        rid = context.id()
        try:
            async for chunk in self._handler.generate(request, context):
                await self._q.put((rid, chunk))
        except asyncio.CancelledError:
            # Client/connection cancellation: let the sentinel close the stream.
            raise
        except Exception as exc:  # noqa: BLE001 - surfaced as a tagged frame
            logging.exception("batched egress feeder failed for request %s", rid)
            await self._q.put(
                (rid, {"finish_reason": {"error": repr(exc)}, "token_ids": []})
            )
        finally:
            await self._q.put((rid, _DONE))

    async def drain(self, context: Optional[Any] = None) -> AsyncGenerator[list, None]:
        """Single multiplexed generator for the whole worker.

        Each ``__anext__`` returns ALL currently-ready ``(rid, chunk)`` pairs as
        one list, so a whole engine step's responses cross the bridge in a single
        crossing. Emits ``list[tuple[str, Optional[dict]]]``; a chunk of ``None``
        marks that request finished (Rust closes its stream and drops the route).
        """
        q = self._q
        while True:
            # Block for the first item so we never spin, then greedily drain
            # everything else already queued into the same batch.
            first_rid, first_chunk = await q.get()
            batch = [(first_rid, None if first_chunk is _DONE else first_chunk)]
            while not q.empty():
                rid, chunk = q.get_nowait()
                batch.append((rid, None if chunk is _DONE else chunk))
            yield batch


async def serve_batched_endpoint(
    endpoint: Any, handler: Any, **serve_kwargs: Any
) -> None:
    """Register the batched-egress engine on ``endpoint``.

    Mirrors ``endpoint.serve_endpoint(handler.generate, ...)`` but routes through
    the Rust ``PythonBatchedEgressEngine`` via ``serve_endpoint_batched``.
    """
    be = BatchedEgress(handler)
    await endpoint.serve_endpoint_batched(be.submit, be.drain, **serve_kwargs)

File: lib/bindings/python/rust/batched_egress.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Cross-request batched egress engine (flag-gated; see DYN-3703).
//!
//! The default `PythonServerStreamingEngine` drives ONE Python async generator
//! per request and crosses the Python->Rust bridge once per token. This engine
//! instead drives a SINGLE multiplexed `drain` generator for the whole worker:
//! each `__anext__` returns a whole engine step's `(request_id, chunk)` pairs,
//! so the batch crosses the bridge in one crossing and is demuxed back to each
//! request's response stream in GIL-free Rust. Analogue of TRT-LLM's
//! `handle_for_ipc_batched`.
//!
//! Python side is `components/src/dynamo/trtllm/request_handlers/batched_egress.py`.
//!
//! BUILD NOTES (this file is a best-effort scaffold; verify at compile):
//!  - `demand_driven_python_stream` and `invoke_generator` in `engine.rs` must
//!    be made `pub(crate)` to reuse them here (or replicate the unfold driver).
//!  - Wire in `lib.rs`: `mod batched_egress;` and an `Endpoint`
//!    `serve_endpoint_batched(submit, drain, ...)` method that builds this
//!    engine + the same `PythonServerStreamingIngress` adapter used by
//!    `serve_endpoint`.
//!  - Confirm the exact `PythonResponseItem` / `PythonPayload` constructors
//!    against `python_payload.rs` (names used below mirror `engine.rs`).

use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};

use anyhow::{Error, Result};
use dashmap::DashMap;
use pyo3::prelude::*;
use pyo3::types::{PyList, PyTuple};
use tokio::sync::mpsc;
use tokio_stream::{StreamExt, wrappers::ReceiverStream};

use dynamo_runtime::pipeline::{
    AsyncEngine, AsyncEngineContextProvider, ManyOut, ResponseStream, SingleIn,
};

use crate::engine::{PythonServerStreamingEngine, demand_driven_python_stream, map_python_exception};
use crate::python_payload::{PythonPayload, PythonResponseItem};
use super::context::Context;

/// Per-request response sink registered while the request is in flight.
type Route = mpsc::Sender<PythonResponseItem>;

const RESPONSE_CHANNEL_DEPTH: usize = 128;

/// Engine that batches egress across requests. Constructed from the Python
/// `submit` coroutine-fn and the `drain` async-generator-fn.
#[derive(Clone)]
pub(crate) struct PythonBatchedEgressEngine {
    submit: Arc<PyObject>,
    drain: Arc<PyObject>,
    event_loop: Arc<PyObject>,
    routes: Arc<DashMap<String, Route>>,
    forwarder_started: Arc<AtomicBool>,
}

impl PythonBatchedEgressEngine {
    pub(crate) fn new(submit: PyObject, drain: PyObject, event_loop: PyObject) -> Self {
        Self {
            submit: Arc::new(submit),
            drain: Arc::new(drain),
            event_loop: Arc::new(event_loop),
            routes: Arc::new(DashMap::new()),
            forwarder_started: Arc::new(AtomicBool::new(false)),
        }
    }

    /// Start the single drain->demux forwarder exactly once.
    fn ensure_forwarder(&self) {
        if self.forwarder_started.swap(true, Ordering::SeqCst) {
            return;
        }
        let drain = self.drain.clone();
        let event_loop = self.event_loop.clone();
        let routes = self.routes.clone();

        tokio::spawn(async move {
            // Build the drain async generator and wrap it as a demand-driven
            // Rust stream (same machinery the per-request path uses).
            let stream = Python::with_gil(|py| -> PyResult<_> {
                let generator = drain.call0(py)?; // drain() -> async generator
                let locals = pyo3_async_runtimes::TaskLocals::new(event_loop.bind(py).clone());
                demand_driven_python_stream(locals, generator.into_bound(py))
            });
            let mut stream = match stream {
                Ok(s) => s,
                Err(e) => {
                    tracing::error!("batched egress: failed to start drain: {e}");
                    return;
                }
            };

            while let Some(item) = stream.next().await {
                let item = match item {
                    Ok(item) => item,
                    Err(e) => {
                        tracing::error!("batched egress drain error: {e}");
                        break;
                    }
                };
                // ONE GIL acquisition converts + splits the whole step's batch
                // into owned (request_id, PythonResponseItem, is_done) triples.
                let routed: Vec<(String, Py<PyAny>, bool)> = Python::with_gil(|py| {
                    let mut out = Vec::new();
                    let list = match item.bind(py).downcast::<PyList>() {
                        Ok(list) => list.clone(),
                        Err(_) => {
                            tracing::error!("batched egress: drain yielded non-list");
                            return out;
                        }
                    };
                    for elem in list.iter() {
                        // each elem is a (request_id, chunk_or_None) tuple
                        let Ok(tup) = elem.downcast::<PyTuple>() else {
                            continue;
                        };
                        let Ok(rid_obj) = tup.get_item(0) else { continue };
                        let Ok(chunk) = tup.get_item(1) else { continue };
                        let Ok(rid) = rid_obj.extract::<String>() else {
                            continue;
                        };
                        let is_done = chunk.is_none();
                        out.push((rid, chunk.unbind(), is_done));
                    }
                    out
                });

                for (rid, chunk, is_done) in routed {
                    if is_done {
                        // Drop the sender -> closes this request's ManyOut.
                        routes.remove(&rid);
                        continue;
                    }
                    if let Some(tx) = routes.get(&rid) {
                        let item = PythonResponseItem::new(Ok(chunk));
                        if tx.send(item).await.is_err() {
                            drop(tx);
                            routes.remove(&rid);
                        }
                    }
                }
            }
        });
    }
}

#[async_trait::async_trait]
impl AsyncEngine<SingleIn<PythonPayload>, ManyOut<PythonResponseItem>, Error>
    for PythonBatchedEgressEngine
{
    async fn generate(
        &self,
        request: SingleIn<PythonPayload>,
    ) -> Result<ManyOut<PythonResponseItem>, Error> {
        self.ensure_forwarder();

        let (request, context) = request.transfer(());
        let ctx = context.context();
        let id = context.id().to_string();

        // Register this request's response route BEFORE submitting so the drain
        // can never deliver a token before the route exists.
        let (tx, rx) = mpsc::channel::<PythonResponseItem>(RESPONSE_CHANNEL_DEPTH);
        self.routes.insert(id.clone(), tx);

        // Schedule submit(request, context) as a fire-and-forget coroutine on the
        // worker's event loop. This starts the per-request feeder that runs the
        // existing handler.generate and pushes tagged chunks onto the shared
        // Python queue that `drain` batches.
        let submit = self.submit.clone();
        let event_loop = self.event_loop.clone();
        let metadata = context.metadata().clone();
        let trace = dynamo_runtime::logging::get_distributed_tracing_context();
        let request_py = request.into_inner();
        let schedule = Python::with_gil(|py| -> PyResult<()> {
            let ctx_obj = Py::new(py, Context::new(ctx.clone(), trace, None, metadata))?;
            let coro = submit.call1(py, (request_py, ctx_obj))?;
            // asyncio.run_coroutine_threadsafe(coro, loop) — fire and forget.
            let asyncio = py.import("asyncio")?;
            asyncio.call_method1(
                "run_coroutine_threadsafe",
                (coro, event_loop.bind(py).clone()),
            )?;
            Ok(())
        });
        if let Err(e) = schedule {
            self.routes.remove(&id);
            return Err(map_python_exception(e).into());
        }

        Ok(ResponseStream::new(
            Box::pin(ReceiverStream::new(rx)),
            context.context(),
        ))
    }
}

/// Constructed by `Endpoint::serve_endpoint_batched` in `lib.rs`.
pub(crate) fn new_engine(
    submit: PyObject,
    drain: PyObject,
    event_loop: PyObject,
) -> PythonBatchedEgressEngine {
    PythonBatchedEgressEngine::new(submit, drain, event_loop)
}

// Silence unused import if the ingress adapter is wired directly in lib.rs.
#[allow(unused_imports)]
use PythonServerStreamingEngine as _KeepEngineImport;
